core/app/models/spree/shipment.rb

```diff
@@ -386,6 +388,33 @@ def manifest_unstock(item)
       stock_location.unstock item.variant, item.quantity, self
     end
 
+    def cancel_inventory_units
+      inventory_units.where.not(state: 'canceled').each(&:cancel!)
+    end
+
+    def resume_inventory_units
+      manifest.each do |item|
+        number_of_items_to_restore = item.states["on_hand"].to_i + item.states["backordered"].to_i
+
+        if item.variant.should_track_inventory?
+          units_available, units_to_back_order = stock_location.fill_status(item.variant, number_of_items_to_restore)
```

Comment:
Are these variables filled with the current availability in the stock location, right?

Reply:
Exactly, this will update the inventory units state based on the current stock level.

This is why inside #after_resume I have to call #resume_inventory_units before unstocking the items. If you call #resume_inventory_units after the unstocking the states of the inventory units will be wrong. 

There is a test covering this use-case to avoid regression.